### DISCUSSION TITLE
Building Mainline kernel

### DISCUSSION THREAD

**Discussion by stevepringle** [Joshua-Riek/ubuntu-rockchip#531] [General]
The Wiki is really helpful in showing how to build the standard rockchip kernel, but doesn't cover building the mainline kernel. What do I need to do to be able to build that myself?

**Comment by bagong** [comment]
The build script in the ubuntu-rockchip repo gives you a lot of options:
````
ubuntu@ubuntu:~/Projects/ubuntu-rockchip$ sudo ./build.sh -h
Usage: ./build.sh --board=[orangepi-5|orangepi-5b|orangepi-5-plus|armsom-w3|armsom-sige7|rock-5b|rock-5a|rock-5-itx|radxa-nx5-io|radxa-cm5-io|nanopc-t6|nanopi-r6c|nanopi-r6s|indiedroid-nova|mixtile-blade3|mixtile-core3588e|lubancat-4|turing-rk1|roc-rk3588s-pc]

Required arguments:
  -b, --board=BOARD      target board 

Optional arguments:
  -h,  --help            show this help message and exit
  -c,  --clean           clean the build directory
  -d,  --docker          use docker to build
  -k,  --kernel-only     only compile the kernel
  -u,  --uboot-only      only compile uboot
  -so, --server-only     only build server image
  -do, --desktop-only    only build desktop image
  -m,  --mainline        use mainline linux sources
  -l,  --launchpad       use kernel and uboot from launchpad repo
  -v,  --verbose         increase the verbosity of the bash script
````
So that's really cool and building mainline is achieved by just typing `sudo ./build.sh -m`. But from an educational point of view my current beef is that a lot of the true magic for using such a kernel is hidden in the call of `dpkg-buildpackage` (in build-u-boot.sh), so you can't really learn what's going on from reading the script. This is not to criticize Joshuas great work, the scripts are very clear and clean, and there is a lot to learn. But they serve the purpose of creating a single kernel-boot-loader-os-package for his splendid Github-actions based build system, and less so the purpose of trying different kernel versions with a given, running system- Well, that's what I understand for now...

**Comment by hakotani-o** [comment]
Please refer to scripts/build-kernel.sh.
